processor/src/stack/overflow.rs

```diff
@@ -1,211 +1,414 @@
 
-    /// Saves a copy of the current table state into the trace at the specified clock cycle.
-    fn save_current_state(&mut self, clk: u64) {
-        debug_assert!(self.trace_enabled, "overflow table trace not enabled");
-        let current_state = self.active_rows.iter().map(|&idx| self.all_rows[idx].val).collect();
-        self.trace.insert(clk, current_state);
+    /// Returns the overflow stack for the current context.
+    ///
+    /// Specifically, this is a reference to the more recent overflow stack in the list of overflow
+    /// stacks for the current context. Recall that for all contexts other than the root context,
+    /// there is at most one overflow stack, but for the root context, there can be two.
+    fn get_current_overflow_stack(&self) -> &OverflowStack {
+        self.overflow
+            .get(&self.current_ctx)
+            .and_then(|overflow_stacks| overflow_stacks.last())
+            .expect("The current context should always have an overflow stack initialized")
     }
 
-    // TEST ACCESSORS
+    /// Mutable version of `get_current_overflow_stack()`.
+    fn get_current_overflow_stack_mut(&mut self) -> &mut OverflowStack {
+        self.overflow
+            .get_mut(&self.current_ctx)
+            .and_then(|overflow_stacks| overflow_stacks.last_mut())
+            .expect("The current context should always have an overflow stack initialized")
+    }
+}
+
+/// Stores the history of the overflow table at every clock cycle, where only the relevant context
+/// is stored in the history for each clock cycle.
+///
+/// The events which update the overflow table history are:
+/// - `push` operation,
+/// - `pop` operation,
+/// - a new context is started,
+/// - a former context is restored.
+#[derive(Debug)]
+struct OverflowTableHistory {
+    /// Stores the full state of the overflow table at every clock cycle. Formally, if `clk <
+    /// last_transition_clk`, then there is exactly one range in the history which contains `clk`.
+    history: Vec<(RangeInclusive<RowIndex>, Vec<Felt>)>,
+}
+
+impl OverflowTableHistory {
+    /// Creates a new empty overflow table history.
+    pub fn new() -> Self {
+        Self { history: Vec::new() }
+    }
+
+    // PUBLIC ACCESSORS
     // --------------------------------------------------------------------------------------------
 
-    #[cfg(test)]
-    pub fn all_rows(&self) -> &[OverflowTableRow] {
-        &self.all_rows
+    /// Returns the overflow table at the given clock cycle in the active context at that clock
+    /// cycle if present in the history, otherwise returns `None`.
+    ///
+    /// That is, if the target clock cycle is greater than the last clock cycle at which the
+    /// history was updated, the method returns `None`.
+    ///
+    /// The first element returned by the iterator is the top of the overflow table.
+    pub fn get_at(&self, target_clk: RowIndex) -> Option<impl Iterator<Item = &Felt>> {
```

Comment:
Related to the previous comment, if we store the state after a given operation, this could be simplified a bit and the signature could be:
```rust
pub fn get_at(&self, target_clk: RowIndex) -> &[Felt]
```

Reply:
We can't return a slice in either case, since we don't store the `Vec<Felt>` in the correct order - we need to `rev()` before returning. So I think the current signature is more flexible, since we can change the internal representation (e.g. by flipping the order of elements stored) without breaking the API. Also, it works well with the `append_into()` pattern that we have (such that we only need to do `result.extend(iterator)`).